### PR TITLE
[model] refactor qwen3.5 GDN to use FLA causal conv1d and cu_seq_lens_q

### PR DESCRIPTION
**Pull request by ys2025-AI**
### Fix

**`src/mcore_bridge/model/mm_gpts/qwen3_5.py`** 

1. **`__init__` hybrid-arch adaptation**: auto-derive `config.layer_types` from `linear_attention_freq` (int → per-layer list); pass `layer_number - 1` to the HF parent (0-based `layer_idx`).
2. **Varlen conv**: override `self.causal_conv1d_fn = self._fla_conv_fn`, which routes FLA `causal_conv1d(..., cu_seqlens=self._cur_cu_seqlens)` — gives the conv sequence-boundary isolation in varlen (solves cross-sample state leakage that `attention_mask` cannot fully prevent on GDN's recurrent state).
3. **Varlen forward**: thd path keeps packed `[1, total, D]` and passes `cu_seq_lens_q` to `super().forward(...)` instead of un-padding to a padded batch + repacking after. Verified the parent `Qwen3_5MoeGatedDeltaNet.forward` (transformers ≥5.14.1, `modeling_qwen3_5_moe.py:546`) consumes `kwargs["cu_seq_lens_q"]` → `chunk_gated_delta_rule(cu_seqlens=...)`, and FLA `causal_conv1d` supports `cu_seqlens` (signature verified), so both conv and delta-rule run varlen.

### CODE REVIEW

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Copyright (c) ModelScope Contributors. All rights reserved.
 import torch
+import torch.nn.functional as F
+from fla.modules.conv.causal_conv1d import causal_conv1d as _fla_causal_conv1d
```
Comment (hjh0119) [inline src/mcore_bridge/model/mm_gpts/qwen3_5.py]: try catch

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py [hunk 2]
```diff
@@ -25,41 +27,48 @@ class Qwen3_5MoeGatedDeltaNet(_HuggingFaceModule, _Qwen3_5MoeGatedDeltaNet):
...
+        hidden_states = hidden_states.transpose(0, 1)
         with get_cuda_rng_tracker().fork('data-parallel-rng'):
-            res = super().forward(hidden_states=hidden_states, attention_mask=attention_mask)
-        if thd_format:
-            res = res[attention_mask][:, None]
-            res = torch.concat([res, res.new_zeros(seq_len - res.shape[0], 1, res.shape[2])])
-        else:
-            res = res.transpose(0, 1).contiguous()
+            res = super().forward(
+                hidden_states=hidden_states, attention_mask=attention_mask, cu_seq_lens_q=cu_seq_lens_q)
```
Comment (hjh0119) [inline src/mcore_bridge/model/mm_gpts/qwen3_5.py]: 确认下支持传入cu_seq_lens_q的transformers版本，低版本可能需要 raise 提示

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py [hunk 3]
```diff
@@ -25,41 +27,48 @@ class Qwen3_5MoeGatedDeltaNet(_HuggingFaceModule, _Qwen3_5MoeGatedDeltaNet):
...
+        if getattr(config, 'layer_types', None) is None:
+            freq = config.linear_attention_freq
+            if isinstance(freq, int):
+                freq = [0 if (i + 1) % freq == 0 else 1 for i in range(config.num_layers)]
+            config.layer_types = ['linear_attention' if f else 'full_attention' for f in freq]
+        _Qwen3_5MoeGatedDeltaNet.__init__(self, config, layer_number - 1)
         self.config = config
         extra_kwargs = _get_extra_te_kwargs(config)
         self.to(dtype=extra_kwargs['params_dtype'], device=extra_kwargs['device'])
+        self.causal_conv1d_fn = self._fla_conv_fn
```
Comment (hjh0119) [inline src/mcore_bridge/model/mm_gpts/qwen3_5.py]: self.causal_conv1d_fn 在 transformers 5.15 删除了

https://github.com/huggingface/transformers/pull/47450

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py (hunk 2 above)
Comment (hjh0119) [reply src/mcore_bridge/model/mm_gpts/qwen3_5.py]: https://github.com/huggingface/transformers/pull/45034 引入

**Review by hjh0119 — CHANGES_REQUESTED** [review]

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py (hunk 1 above)
Comment (ys2025-AI) [reply src/mcore_bridge/model/mm_gpts/qwen3_5.py]: 已将 fla 导入用 try/except ImportError 包裹，并设置 _fla_causal_conv1d = None

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py (hunk 2 above)
Comment (ys2025-AI) [reply src/mcore_bridge/model/mm_gpts/qwen3_5.py]: 已确认 — cu_seq_lens_q 是在 transformers v5.9.0 中引入的。在 thd 路径中增加了版本检查，当 <5.9.0 时抛出 RuntimeError，并提示所需的版本和当前的 transformers.__version__。仅在 cu_seq_lens_q 为非 None 时（即 thd 路径 ≥ 5.9.0）将其转发给 super().forward()，因此非 packed 的前向传播在旧版本上依然有效。

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py (hunk 3 above)
Comment (ys2025-AI) [reply src/mcore_bridge/model/mm_gpts/qwen3_5.py]: 是的，已确认causal_conv1d_fn 在 transformers v5.15.0 中被移除。已添加版本检查限制：if _fla_causal_conv1d is not None and _TRANSFORMERS_HAS_CONV1D_FN，其中 _TRANSFORMERS_HAS_CONV1D_FN = version.parse(transformers.__version__) < version.parse('5.15.0')。这与 requirements.txt 中现有的 transformers<5.15.0 上限一致，且不会在 5.15+ 版本中崩溃。

File: src/mcore_bridge/model/mm_gpts/qwen3_5.py [hunk 4]
```diff
@@ -19,47 +22,75 @@
...
-        if thd_format:
-            res = res[attention_mask][:, None]
-            res = torch.concat([res, res.new_zeros(seq_len - res.shape[0], 1, res.shape[2])])
-        else:
-            res = res.transpose(0, 1).contiguous()
+            if cu_seq_lens_q is not None:
+                res = super().forward(
+                    hidden_states=hidden_states, attention_mask=attention_mask, cu_seq_lens_q=cu_seq_lens_q)
+            else:
+                res = super().forward(hidden_states=hidden_states, attention_mask=attention_mask)
```
Comment (hjh0119) [inline src/mcore_bridge/model/mm_gpts/qwen3_5.py:92]: else 是不是会跨序列泄漏，后续支持 5.15+ 需要额外支持吗

Comment (ys2025-AI) [reply src/mcore_bridge/model/mm_gpts/qwen3_5.py:92]: @hjh0119 不会跨序列泄漏。else 分支走的是标准非 packed 路径，输入为 (batch, seq, hidden) 格式，batch 维度天然隔离样本，配合 attention_mask 处理 padding 和因果掩码，不存在泄漏问题。不过 transformers 5.15+ 确实需要额外适配。因为重构了 GDN 的卷积路径，causal_conv1d_fn 的调用签名发生变化（activation 从 keyword 参数改为 positional），当前 _fla_conv_fn 的注入机制在 5.15+ 下会失效。变长序列场景需要重新验证 cu_seqlens 的传递和卷积隔离正确性。

**Review by hjh0119 — APPROVED** [review]

**Comment by hjh0119** [comment]
thanks, can you resolve the conflicts?

**Comment by ys2025-AI** [comment]
> thanks, can you resolve the conflicts?

@hjh0119 Conflicts resolved, ready for another look. Thanks!

**Review by hjh0119 — APPROVED** [review]
